src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
+			zap.Int("numShards", len(currNamespace.DataRunOptions.ShardTimeRanges)),
+			zap.Duration("dataRange", dataRange),
+		}...)
+		if dataRange > 0 {
+			dataMin, dataMax := currNamespace.DataRunOptions.ShardTimeRanges.MinMax()
+			prepareLogFields = append(prepareLogFields, []zapcore.Field{
+				zap.Time("dataFrom", dataMin),
+				zap.Time("dataTo", dataMax),
+			}...)
+		}
+
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			indexRange := currNamespace.IndexRunOptions.ShardTimeRanges.MinMaxRange()
+			prepareLogFields = append(prepareLogFields, []zapcore.Field{
+				zap.Duration("indexRange", indexRange),
+			}...)
+			if indexRange > 0 {
+				indexMin, indexMax := currNamespace.IndexRunOptions.ShardTimeRanges.MinMax()
+				prepareLogFields = append(prepareLogFields, []zapcore.Field{
+					zap.Time("indexFrom", indexMin),
+					zap.Time("indexTo", indexMax),
+				}...)
+			}
+		}
+
+		b.log.Info("bootstrap from source for namespace ready", prepareLogFields...)
 	}
-	b.log.Info("bootstrapping from source starting", logFields...)
 
 	nowFn := b.opts.ClockOptions().NowFn()
 	begin := nowFn()
 
-	currStatus, currErr = step.runCurrStep(currRanges)
+	b.log.Info("bootstrap from source started", logFields...)
+	currResults, err := b.src.Read(curr)
 
 	logFields = append(logFields, zap.Duration("took", nowFn().Sub(begin)))
-	if currErr != nil {
-		logFields = append(logFields, zap.Error(currErr))
-		b.log.Info("bootstrapping from source completed with error", logFields...)
-	} else {
-		logFields = append(logFields, currStatus.logFields...)
-		b.log.Info("bootstrapping from source completed successfully", logFields...)
+	if err != nil {
+		errorLogFields := append(logFieldsCopy(logFields), zap.Error(err))
+		b.log.Info("bootstrapping from source completed with error", errorLogFields...)
+	}
+	if err != nil {
+		return bootstrap.NamespaceResults{}, err
 	}
 
-	wg.Wait()
-	if err := xerrors.FirstError(currErr, nextErr); err != nil {
-		return err
+	// Determine the unfulfilled and the unattempted ranges to execute next.
+	next := bootstrap.Namespaces{
+		Namespaces: bootstrap.NewNamespacesMap(bootstrap.NamespacesMapOptions{}),
 	}
+	for _, elem := range namespaces.Namespaces.Iter() {
+		id := elem.Key()
+		namespace := elem.Value()
+
+		currResult, ok := currResults.Results.Get(id)
+		if !ok {
+			return bootstrap.NamespaceResults{},
+				fmt.Errorf("namespace result not returned by bootstrapper: %v", id.String())
+		}
+
+		currNamespace, ok := curr.Namespaces.Get(id)
+		if !ok {
+			return bootstrap.NamespaceResults{},
+				fmt.Errorf("namespace prepared request not found: %v", id.String())
+		}
+
+		// Shallow copy the current namespace for the next namespace prepared request.
+		nextNamespace := currNamespace
+
+		// Calculate bootstrap time ranges.
+		dataRequired := namespace.DataRunOptions.ShardTimeRanges.Copy()
+		dataCurrRequested := currNamespace.DataRunOptions.ShardTimeRanges.Copy()
+		dataCurrFulfilled := dataCurrRequested.Copy()
```

Comment:
nit: don't think this copy is necessary

Reply:
True, I'm going to opt to just copy everywhere though if that's ok for this code path since it only executes less than 3 times each bootstrap, and easier to not think and just copy it at every single place for safety.